### DISCUSSION TITLE
Extending the USBMidiKliK4x4?

### DISCUSSION THREAD

**Discussion by TonyRoe** [TheKikGen/USBMidiKliK4x4#52] [Q&A]
Hi. I'm new here (from a long hardware background) and also to embedded programming. I've made some progress on a digital audio mixer (there are some great codecs available now) and on two MIDI synths - one based on a partial Yamaha SY77, and one a new concept in guitar synthesis which efficiently models things like picking styles and positions - all in a separate "audio engine". I intended to add virtually everything that USBMidiKliK4x4 does, but it now seems much more logical to just use USBMidiKliK4x4 virtually as-is, but also sending MIDI streams to the audio engine - maybe extras, or maybe replacing either one of the 3+3 slave modules or some of the internal virtual ports if the port numbers are capped? I would need at least two full MIDI streams including sysex to control the synths and mixer (maybe unused commands in the USBMidiKliK4x4 sysex space?). I'm OK coding my own little isolated things, but so far I haven't made much progress understanding the whole USBMidiKliK4x4 project code, so please forgive my ignorance. Any guidance would be appreciated.

**Comment by TheKikGen** [comment]
Hi Tony. I 've just seen your post.
You can send sysex already on any port. 
Tell me more with examples.

About documentation, you need to read the code it self.  it is commented.
Start with the main ino file.

I started a "midi chord" implementation here : https://github.com/TheKikGen/USBMidiKliK4x4/blob/54db451fa74b1209ed67c07f733191f31ce77520/mod_miditransfn_pipes.h#L857

This is maybe a good base for you to add a transformation pipe.

**Comment by TheKikGen** [comment]
I'm interrested in  your "new concept in guitar synthesis which efficiently models things like picking styles and positions".
USBMDK is able to manage 3 IN/ 3OUT physical ports, and 16 IN/OUT usb ports.
So not a problem at all to dedicated IN/OU ports to sysex commands.
You could use port 1 In/Out for SYSEX and port 2 for midi notes for example.

**Reply by TonyRoe** [reply]
Hi. Thank you for your interest.
I'm not sure I'll be able to continue the project as I've had several strokes with a lot of cognitive impairment. But the basic idea is this:
Each guitar pickup pole set, the pick location, and the geometry of the two active string segments DURING pick contact time all need comb filters. Instead of subtractively filtering a wrong wave, an initial multi-stepped rectangular waveform can be quickly calculated once, then generated with ALL needed harmonics in the correct polarities, of all pickup pole sets, balanced as needed (even allowing the energy in the pickups' "null plane" (which still generates a little octave-up audio) to morph into the most active plane during the decay phase.
Only a few actions are then needed per fundamental frequency cycle (depending on pickup complexity), rather than on every audio sample. Each waveform step comprises two partial steps, for precise sub-sample timing accuracy.
Each string's audio then has an adaptive variable LPF' and comb filter for the pick position; picking divides the string into two separate sections, with amplitude decaying quickly (for clean country picking) to not (if the pick is dragged across the string).; the discord of the two musically unrelated notes is too brief to measure, but creates most of the pick sound; some noise might also need to be added.
MIDI notes can be on 6 separate MIDI channels (e.g. from a DAW or hex MIDI pickup; best for pedal steel-like independent pitchbend, or any other bend leaving some strings unaffected), or guessed by an algorithm, or shifted to all fit into one MIDI channel (considering the higher frets tend to be used only for the higher strings, e.g. allowing 24/24/22/20/19/18 frets for the EBGDAE strings); in that case, the bottom 3 bits of pitchbend values also determine if the bend should be applied to all strings, or just to one of strings 1 to 7 (max error < 1 cent). All the rest is pretty standard subtractive synth stuff, but simpler as the initial waveform does the heavy lifting, and memory accesses are minimized. I had similar plans for an ultra-DX7 and a generic digital mixer, all on one core of a Raspberry Pi (or Odroid etc).
